transport_plugins/jlink/iotile_transport_jlink/jlink_background.py

```diff
@@ -1,188 +1,113 @@
-    }
-
-    def __init__(self, jlink):
-        super(JLinkControlThread, self).__init__()
-
-        self._jlink = jlink
-        self._commands = Queue()
-
-    def run(self):
-        logger.critical("Starting JLink control thread")
-        while True:
-            try:
-                cmd = self._commands.get()
-                if not isinstance(cmd, JLinkCommand):
-                    logger.error("Invalid command object that is not a JLinkCommand: %s", cmd)
-                    continue
-
-                if cmd.name == JLinkControlThread.STOP:
-                    logger.info("stop command received")
-                    break
-
-                callback = cmd.callback
-                exception = None
-                result = None
-                try:
-                    funcname = self.KNOWN_COMMANDS.get(cmd.name)
-                    if funcname is None:
-                        raise ArgumentError("Unknown command name in JLinkControlThread", name=cmd.name)
-
-                    func = getattr(self, funcname)
-
-                    args = cmd.args
-                    if args is None:
-                        args = ()
-
-                    result = func(*args)
-                except Exception as exc:  #pylint:disable=broad-except;We want to rethrow everything in the main thread
-                    exception = exc
-                    logger.exception("Error executing command %s with args %s", cmd.name, cmd.args)
-
-                if callback is not None:
-                    callback(cmd.name, result, exception=exception)
-            except:  #pylint:disable=bare-except;We want to keep our control thread always open
-                logger.exception("Uncaught exception in JLink control thread")
-
-        logger.critical("Shutting down JLink control thread")
-
-    def command(self, cmd_name, callback, *args):
-        """Run an asynchronous command.
-
-        Args:
-            cmd_name (int): The unique code for the command to execute.
-            callback (callable): The optional callback to run when the command finishes.
-                The signature should be callback(cmd_name, result, exception)
-            *args: Any arguments that are passed to the underlying command handler
-        """
-        cmd = JLinkCommand(cmd_name, args, callback)
-        self._commands.put(cmd)
+    def __init__(self, jlink_adapter, loop=SharedLoop):
+        self._jlink_adapter = jlink_adapter
+        self._jlink = None
+        self._loop = loop
 
-    def stop(self):
-        """Tell this thread to stop, do not wait for it to finish."""
-        self._commands.put(JLinkCommand(JLinkControlThread.STOP, None, None))
+        # maitenance can be enabled for different interfaces
+        self._maitenance_counter = 0
+        self._maitenance_task = None
+        self.rpc_response_event = asyncio.Event() # triggered when the controller acknowledge rpc response
```

Comment:
All `asyncio` primitives like `Event`, `Lock`, etc need to be created by requesting from the `BackgroundEventLoop` passed in.  If you request directly from `asyncio` it will create a different `asyncio` loop and attach the event to that.

You can see this if you run `iotile` tool with `-vvvvv` option and look at the logs.  You should only ever see **one line** that says something like "Starting asyncio event loop using method SelectorSomething".  If you see multiple lines like that it means that you're asyncio primitives are not all attached to the same loop and the code **will break** in subtle and hard to diagnose ways.

The safe way to get things from `asyncio` is to request them through the `loop` object you were passed in to ensure you don't accidentally create another loop.

Reply:
It's my bad, forget about background loop.